<a href="https://colab.research.google.com/github/shaliha223/DSA_class_activities/blob/main/case_study_stati_shaliha.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1. Dataset: Monthly temperatures in a city (in °C): 12, 15, 14, 16, 18, 20, 22, 19,
17, 15, 13, 10 Questions:
a) Calculate the mean temperature.
b) Calculate the median temperature.
c) Find the mode of the temperature data.
d) Range and Variance

In [1]:
Temperature = [12, 15, 14, 16, 18, 20, 22, 19, 17, 15, 13, 10]

def mean(x):
  m = 0

  for i in x:
    m+= i
  mean = m / len(x)

  return mean

print("mean temperature:",mean(Temperature))

mean temperature: 15.916666666666666


In [ ]:
def median(x):

  x.sort()
  l = len(x)

  if l % 2 != 0:
    mid = l // 2
    median = x[mid]

  else:
    mid = l // 2
    median = (x[mid - 1] + x[mid]) / 2

  return median

print("median of temperature:",median(Temperature))

median of temperature: 15.5


In [ ]:
def mode(x):

  count={}

  for i in x:
    if i in count:
      count[i]+=1

    else:
      count[i]=1


  max_count=max(count.values())

  if max_count == 1:
    mode = None

  else:
    mode = []

    for a, b in count.items():
      if b == max_count:
          mode.append(a)

  return mode

print("mode of the temperature:", mode(Temperature))

mode of the temperature: [15]


In [3]:
def range(x):

  range = max(Temperature) - min(Temperature)

  return range

print("range of the temperature:", range(Temperature))

def variance(x):

  m = 0
  v=0

  for i in x:
    m+= i

  mean = m / len(x)

  for i in x:
    v+=((i-mean)**2)
  variance=v/len(x)

  return variance

print("variance of the temperature:",variance(Temperature))

range of the temperature: 12
variance of the temperature: 11.076388888888888


In [4]:
def standard_deviation(x):

  s_d=variance(x)**(1/2)

  return s_d

2. Dataset: Heights of students (in cm): 150, 160, 165, 170, 155, 160, 162, 168
Questions: a) Calculate the range of the heights.
b) Calculate the variance of the heights.
c) Standard Deviation

In [ ]:
Height = [150, 160, 165, 170, 155, 160, 162, 168 ]

print("range of the Heights:",range(Height))
print("variance of the heights:",variance(Height))
print("standard deviation :",standard_deviation(Height))

range of the Heights: 12
variance of the heights: 38.1875
standard deviation : 6.179603547154137


3. Dataset: Weekly expenses (in $): 200, 250, 300, 220, 270, 290, 310
Questions: a) Calculate the standard deviation of the weekly expenses.


In [ ]:
we=[ 200, 250, 300, 220, 270, 290, 310]

print("standard deviation of the weekly expenses:",standard_deviation(we))

standard deviation of the weekly expenses: 38.438925848782034


4. Dataset: Advertising Spending (in $1000s): 10, 20, 15, 25, 30; Sales (in
$1000s): 100, 150, 120, 180, 200
Questions: a) Calculate the covariance between advertising spending and sales.
b) Correlation

In [9]:
def pearson_coefficient(x, y):

  cov=0

  for a, b in enumerate(x):

    for c, d in enumerate(y):

      if a == c:

        cov+= (b - mean(x))*(d - mean(y))

  covariance = cov / len(x)
  p_c_c = covariance /(standard_deviation(x)*standard_deviation(y))

  return covariance, p_c_c


In [ ]:
AS = [10,20,15,25,30]
S = [100, 150, 120, 180, 200]

print("covariance between advertising and sales:",pearson_coefficient(AS,S)[0])
print("correlation:",pearson_coefficient(AS,S)[1])

covariance between advertising and sales: 260.0
correlation: 0.9970544855015816


5. Dataset: Study Hours: 2, 4, 6, 8, 10; Exam Scores: 55, 60, 65, 70, 75
Questions: a) Calculate the correlation coefficient between study hours and exam
scores.
b) Interpret the result.
c) Quartiles and IQR

In [10]:
SH = [2, 4, 6, 8, 10]
S = [55, 60, 65, 70, 75 ]

print("correlation coeffcient:",pearson_coefficient(SH,S)[1])

correlation coeffcient: 0.9999999999999998


the correlatioon coefficient between study hour and score are 0.9999 it is  very close to 1 so the correlation is highly positively correlated

In [11]:
import statistics as stats


q1_sh = stats.quantiles(SH, n=4)[0]
q2_sh = stats.median(SH)
q3_sh = stats.quantiles(SH, n=4)[2]
IQR_sh = q3_sh - q1_sh
print("Quartiles and IQR of study hours: Q1:", q1_sh, "Q2:", q2_sh, "Q3:", q3_sh,"IQR:",IQR_sh)

q1_s = stats.quantiles(S, n=4)[0]
q2_s = stats.median(S)
q3_s = stats.quantiles(S, n=4)[2]
IQR_s = q3_s - q1_s
print("Quartiles and IQR of scores: Q1:", q1_s, "Q2:", q2_s, "Q3:", q3_s,"IQR:",IQR_s)

Quartiles and IQR of study hours: Q1: 3.0 Q2: 6 Q3: 9.0 IQR: 6.0
Quartiles and IQR of scores: Q1: 57.5 Q2: 65 Q3: 72.5 IQR: 15.0


6. Dataset: Monthly savings (in $): 200, 300, 400, 500, 600, 700, 800, 900, 1000,
1100 Questions: a) Calculate the first quartile (Q1) and third quartile (Q3).
b) Calculate the interquartile range (IQR).
c) Detecting Outliers


In [12]:
ms = [200, 300, 400, 500, 600, 700, 800, 900, 1000, 1100]

q1 = stats.quantiles(ms, n=4)[0]
q3 = stats.quantiles(ms, n=4)[2]

print("First quartile (Q1) and third quartile (Q3)",q1 ,"and",q3)

First quartile (Q1) and third quartile (Q3) 375.0 and 925.0


In [ ]:
def outliers(x):

  outliers=[]

  x.sort()
  l = len(x)
  Q1 = x[int(l*0.25) - 1]
  Q3 = x[int(l*0.75)-1]
  IQR = Q3 -Q1
  lower_limit = Q1 - IQR*1.5
  upper_limit = Q3 + 1.5*IQR

  for i in x:

    if i >upper_limit or i < lower_limit:
       outliers.append(i)

  return outliers


In [ ]:
IQR = q3 -q1
print("IQR:",IQR)

IQR: 550.0


In [ ]:
outliers(ms)

[]

7. Dataset: Ages of employees: 22, 25, 28, 30, 35, 40, 45, 50, 55, 100
Questions: a) Identify any outliers using the IQR method.


In [ ]:
Age = [22, 25, 28, 30, 35, 40, 45, 50, 55, 100 ]

outliers(Age)

[100]

10. Dataset: Annual salaries of employees (in $1000s): 40, 42, 45, 47, 50, 52, 55,
57, 60, 62 Questions: a) Calculate the z-scores for each salary.
b) Interpret the z-scores.
c) Normal Distribution

In [ ]:
def z_score(x):

  z=[]
  m=0
  v=0

  for i in x:
    m+= i
  mean = m / len(x)

  for i in x:
    v+=((i-mean)**2)

  variance=v/len(x)
  s_d=variance**(1/2)


  for i in x:
    Z = (i-mean) / s_d
    z.append(Z)

  return z


In [13]:
l = [40, 42, 45, 47, 50, 52, 55, 57, 60, 62]

In [14]:
mean(l)

51.0

In [15]:
standard_deviation(l)

7.14142842854285

In [ ]:
z_score(l)

[-1.5403080924308108,
 -1.2602520756252087,
 -0.8401680504168059,
 -0.5601120336112039,
 -0.14002800840280097,
 0.14002800840280097,
 0.5601120336112039,
 0.8401680504168059,
 1.2602520756252087,
 1.5403080924308108]



*   A negative z-score means the salary is below the mean.here 40k has z = -1.54, meaning it is 1.54 standard deviations below average.
*   A positive z-score means the salary is above the mean.Here 62k has z = 1.54, meaning it is 1.54 standard deviations above average.
*  Salaries close to 51k have z-scores near 0 (e.g., 50k = -0.14, 52k = 0.14).

The distribution of the given salary data is approximately normal because, the data is symmetrically spread around the mean salary of 51 ($1000s),the z-scores range between -1.54 and +1.54.In a normal distribution, about 95% of data lies within ±2σ and 99.7% within ±3σ. Here, all values fall inside this range, so there are no extreme outliers.so the shape of the distribution is bell-like

8. Dataset: Observed frequencies of customer preferences for three products:
Product A: 50, 55, 60
Product B: 30, 25, 35
Product C: 20, 20, 25
a)  Perform a chi-square test to determine if there is a significant
difference in customer preferences among the three products.  
b)   Hypothesis Testing




In [2]:
l=[[50, 55, 60],
   [30, 25, 35],
   [20, 20, 25]]
a=0
c=0
row_sum=[]
for i in l:
  for j in i:
    c=c+j
print(c)
for i in l:
  for j in i:
    a=a+j
  row_sum.append(a)
  a=0
print(row_sum)
s=0
col_sums=[]
for col in range(len(l[0])):   # loop over columns
    s = 0
    for row in range(len(l)):  # loop over rows
        s= s + l[row][col]
    col_sums.append(s)

print("Column sums:", col_sums)
p=[]
for i in row_sum:
  E=[]
  for j in col_sums:
    e=i*j/c
    E.append(e)
  p.append(E)
print(p)
g=0
for i in range (len(l)):
  for j in range (len(p)):
    o=l[i][j]
    e=p[i][j]
    chi=(o-e)**2/e
    chi_sq=g+chi
print("Chi square:",chi_sq)
print("Degree of freedom",(len(row_sum)-1)*(len(col_sums)-1))
critical_value=9.488 #from chi-square table df=4 and alpha=0.05
if chi_sq>critical_value:
  print("Reject H0 (No difference)")
else:
  print("Accept H0: there is no difference")

320
[165, 90, 65]
Column sums: [100, 100, 120]
[[51.5625, 51.5625, 61.875], [28.125, 28.125, 33.75], [20.3125, 20.3125, 24.375]]
Chi square: 0.016025641025641024
Degree of freedom 4
Accept H0: there is no difference


9. Dataset: Sample weights (in kg) of a new diet program: 70, 68, 72, 71, 69, 73,
72, 74, 70, 71 Questions: a) Test the hypothesis that the mean weight after the
diet program is equal to 70 kg at a 5% significance level. b) Z-Scores

In [ ]:
A = [70, 68, 72, 71, 69, 73, 72, 74, 70, 71]
Z = z_score(A)
print(Z)

[-0.5773502691896258, -1.7320508075688774, 0.5773502691896258, 0.0, -1.1547005383792517, 1.1547005383792517, 0.5773502691896258, 1.7320508075688774, -0.5773502691896258, 0.0]


In [ ]:
m = mean(A)
sd = standard_deviation(A)
n = len(A)
mu0 = 70
z = (m - mu0) / (sd / (n**0.5))

print("Sample mean:", m)
print("Sample std dev:", sd)
print("Z statistic:", z)

# Decision
if abs(z) > 1.96:  # critical value at 5% two-tailed
    print("Reject H0: significant difference")
else:
    print("Fail to reject H0: no significant difference")

Sample mean: 71.0
Sample std dev: 1.7320508075688772
Z statistic: 1.8257418583505538
Fail to reject H0: no significant difference


13. Dataset: Monthly income of individuals (in $1000s): 3.5, 4.0, 4.2, 4.5, 4.7, 5.0,5.3, 5.5, 5.7, 6.0
Questions: a) Calculate the 95% confidence interval for the mean monthly
income

In [ ]:
import scipy.stats as stats

B = [3.5, 4.0, 4.2, 4.5, 4.7, 5.0,5.3, 5.5, 5.7, 6.0 ]
m = mean(B)
sd = standard_deviation(B)
n = len(B)
alpha = 0.05
t_critical = stats.t.ppf(1 - alpha/2, df=n-1)
margin_of_error = t_critical * (sd/ (n**0.5))
lower_bound = m - margin_of_error
upper_bound = m + margin_of_error

print(f"Sample Mean = {m:.2f}")
print(f"95% Confidence Interval = ({lower_bound:.2f}, {upper_bound:.2f})")

Sample Mean = 4.84
95% Confidence Interval = (4.30, 5.38)


12. Dataset: Sample data of a new drug's effect on blood pressure reduction (in
mmHg): 10, 12, 14, 15, 13, 16, 18, 17, 19, 20
Questions: a) Perform a t-test to determine if the new drug has a significant effect
on blood pressure reduction. b) Report the p-value and interpret the results. c)
Confidence Intervals

In [7]:

from scipy import stats
import numpy as np

# Data (blood pressure reduction in mmHg)
data = [10, 12, 14, 15, 13, 16, 18, 17, 19, 20]

# Hypothesized population mean (H0: mean = 0, no reduction)
# Since we are checking if drug has an effect, we compare against 0
t_stat, p_value = stats.ttest_1samp(data, popmean=0)

# Calculate 95% Confidence Interval for the mean
n = len(data)
m = mean(data)
sd = standard_deviation(data)
se = sd / (n**0.5) # Standard error of mean
ci = stats.t.interval(0.95, df=n-1, loc=m, scale=se)

print("Sample Mean:", m)
print("t-statistic:", t_stat)
print("p-value:", p_value)
print(f"95% Confidence Interval: ({ci[0]:.2f}, {ci[1]:.2f})")



Sample Mean: 15.4
t-statistic: 15.198684153570664
p-value: 1.0062192292419559e-07
95% Confidence Interval: (13.23, 17.57)


The p value is very close to 0 (because p value is 1.006*10^-7)This value is much smaller than significance level 0.05.so reject reject the null hypothesis
𝐻0:𝜇=0.It means the drug has a highly significant effect on reducing blood pressure.